acceptance/suites/tests/555-puppetserver-service/reboot.rb

```diff
@@ -0,0 +1,55 @@
+if (options[:type] == 'pe') then
+  pssn='pe-puppetserver'
+else
+  pssn='puppetserver'
+end
+
+test_name "Validates the ability of the #{pssn} service to reliably start after a reboot."
+
+max=3
+sleeptime=45
+retryCounter=0
+step "\n-----------\nValidate #{pssn} is up and running, then reboot, repeat #{max} times\n-----------"
+
+for i in 1..max 
+  step "\n-----\nLoop #{i}\n-----"
+   
+  step "check uptime"
+  origUptime=on(master, "awk '{print $1}' /proc/uptime").stdout.chomp.to_i
+
+  step "reboot"
+  #rebooting with on(master,'reboot') causes an IOError: stream end.  This seems like a sane workaround.
+  #QENG-1914 for more...
+  system "ssh -i #{options[:keyfile]} -o StrictHostKeyChecking=no root@#{master.reachable_name} \'reboot\'"
+
+  step "sleeping for 15 seconds and waiting for reboot."
+  sleep 15
+
+  unless port_open_within?(master,22,120)
+    raise Beaker::DSL::FailTest, 'Could not SSH into puppet master within 2 minutes after reboot'
+  end
+
+  step "look at uptime to ensure that we actually rebooted"
+  secondUptime=on(master, "awk '{print $1}' /proc/uptime").stdout.chomp.to_i
+  assert_operator(origUptime, :>, secondUptime, "Expect original uptime #{secondUptime} to be greater than original uptime #{origUptime} ")
+  
+  step "sleeping for #{sleeptime} seconds and waiting for #{pssn} to start."
+  sleep sleeptime
+
+  step "Check to see that #{pssn} is up and running."
+  ec = on(master,"service #{pssn} status", :acceptable_exit_codes => 0..99).exit_code
```

Comment:
The `0.99` makes a lot more sense now that I understand the script is trying to first detect that the service has started, then check to see later that it is actually 0.

This seems like a perfect place to try and figure out a solution to the 3-state problem of first waiting until the service is ready to start, then try and start the service, then wait to see if the service successfully started.

Would it be possible to check something else, e.g. the log file for a "starting up!" message for the first transition, then block again on the "ready to service requests!" event before the final service status check?

We've punted on this issue previously by employing polling loops against the TCP port, but this is definitely another situation where it'd be great to try and clearly define the three states of not-yet-started, started-but-not-yet-ready, and fully-operational-deathstar, with affordances for blocking until a transition to the next state transition happens.